src/gui/torrentcontentmodel.cpp

```diff
@@ -467,64 +506,286 @@ int TorrentContentModel::rowCount(const QModelIndex &parent) const
     if (parent.column() > 0)
         return 0;
 
-    TorrentContentModelFolder *parentItem;
+    const TorrentContentModelFolder *parentItem;
     if (!parent.isValid())
         parentItem = m_rootItem;
     else
-        parentItem = dynamic_cast<TorrentContentModelFolder*>(static_cast<TorrentContentModelItem*>(parent.internalPointer()));
+        parentItem = dynamic_cast<const TorrentContentModelFolder*>(this->citem(parent));
 
     return parentItem ? parentItem->childCount() : 0;
 }
 
+// Mime data binary format: File index, followed by stub of path that will be preserved on rename
+QMimeData *TorrentContentModel::mimeData(const QModelIndexList &indexes) const
+{
+    // folders take precedence over files in the sense that if a folder and some files inside of it
+    // are selected, the whole folder is moved and the folder structure is preserved.
+    QHash<int, QString> stubs;
+    QSet<QString> foldersProcessed;
+    std::function<void (const TorrentContentModelItem *, const QString &)> addStubs;
+    addStubs = [&stubs, &foldersProcessed, &addStubs](const TorrentContentModelItem *item, const QString &stubSoFar)
+        {
+            switch (item->itemType())
+            {
+            case TorrentContentModelItem::ItemType::FileType:
+            {
+                // don't override existing put there by folders
+                int fileIndex = dynamic_cast<const TorrentContentModelFile *>(item)->fileIndex();
+                if (!stubSoFar.isEmpty() || !stubs.contains(fileIndex))
+                {
+                    stubs.insert(fileIndex, Utils::Fs::combinePaths(stubSoFar, item->name()));
+                }
+                break;
+            }
+            case TorrentContentModelItem::ItemType::FolderType:
+            {
+                // ok to override existing put there by files
+                // if any parent folders have been processed, they take precedence
+                for (const QString &parent : Utils::Fs::parentFolders(item->path()))
```

Comment:
`const QVector<QString> parentFolders = Utils::Fs::parentFolders(item->path());`
and use that container here

or

```suggestion
                for (const QString &parent : asConst(Utils::Fs::parentFolders(item->path())))
```